Feature/api add invitation get method

ivr42 opened the pull request:
Get all Invitations feature has been added

 - `api/response_models/administrator_invitation.py`: file was created with `AdministratorInvitationResponse` model
 - `api/routers/administrator_invitation.py`: endpoint `/invite` was changed to `/invitations` in `post` decorator
 - `api/routers/administrator_invitation.py`: method `get_all_invitations` was added
 - `core/services/administrator_invitation.py`: method `list_all_invitations` was added
 - Refactored: `async def get_all` added to `AbstractRepository`
 - Refactored: `get_tasks` removed from `TaskRepository`, utilized `get_all` instead
 - Refactored: `expired_date` renamed to `expired_datetime` in the `AdministratorInvitation` model and through all code 

Ссылка на задачу в Notion:
https://www.notion.so/API-51d6e6ed8ec3465ba23514e05156fa0d?pvs=4

ivr42 committed:
Get all Invitations feature has been added

 - `api/response_models/administrator_invitation.py`: file was created with AdministratorInvitationResponse model
 - `api/routers/administrator_invitation.py`: endpoint '/invite' was changed to '/invitations' in router.post
 - `api/routers/administrator_invitation.py`: method `get_all_invitations` was added
 - `core/services/administrator_invitation.py`: method `list_all_invitations` was added
 - `core/db/repository/administrator_invitation.py`: method `get_invitaions` was added

ivr42 committed:
Merge branch 'develop' into feature/api_add_invitation_get_method

ADesBiysk commented on src/api/response_models/administrator_invitation.py:
```diff
@@ -0,0 +1,15 @@
+    name: str
+    surname: str
+    email: str
+    role: str = Administrator.Role.PSYCHOLOGIST
```
По заданию нужно возвращать роль приглашенного, но это не имеет смысла, поскольку у всех она будет одинаковая. А вот время окончания действия нужно возвращать, чтобы администратор мог видеть какие приглашения ещё активные, а какие уже "протухли".

ivr42 replied:
Поле `role` убрал, добавил поле `expired_datetime`

ADesBiysk commented on src/core/db/repository/administrator_invitation.py:
```diff
@@ -23,3 +23,23 @@ async def get_mail_request_by_token(self, token: str) -> Optional[AdministratorI
             raise AdministratorInvitationInvalid
         return result
+
+    async def get_invitaions(self) -> list[AdministratorInvitation]:
```
Не информативное имя.

ivr42 replied:
Переименовал `get_invitaions` в `get_all_invitations`

ADesBiysk commented on src/core/db/repository/administrator_invitation.py:
```diff
@@ -23,3 +23,23 @@ async def get_mail_request_by_token(self, token: str) -> Optional[AdministratorI
+    async def get_invitaions(self) -> list[AdministratorInvitation]:
+
+        result_statement = (
+            select(
```
Не нужно так сложно. По заданию требуется просто вернуть список приглашений. Не требуется выполнять связь с Administrator.

ivr42 replied:
Связь с `Administrator` убрал, сделал проще

ivr42 committed:
Code review fixes:

 - removed the `role` field from API response
 - added the `expired_datetime` to API response
 - refactored: method get_invitaions renamed to get_all_invitations

ivr42 committed:
Merge branch 'develop' into feature/api_add_invitation_get_method

ADesBiysk commented on src/api/routers/administrator_invitation.py:
```diff
@@ -29,3 +33,13 @@ async def create_and_send_invitation(
+        response_model=list[AdministratorInvitationResponse],
+        status_code=HTTPStatus.OK,
+        summary="Получить информацию о приглашениях, отправленных администраторам",
+        response_description="Информация о приглашениях",
```
В этот метод и метод выше нужно добавить аттрибут responses со всеми возможными кодами ответов, кроме OK. Посмотреть образец можно в других эндпойтнах, например PATCH /reports/{report_id}/approve.

ADesBiysk commented on src/api/routers/administrator_invitation.py:
```diff
@@ -29,3 +33,13 @@ async def create_and_send_invitation(
+        summary="Получить информацию о приглашениях, отправленных администраторам",
+        response_description="Информация о приглашениях",
+    )
+    async def get_all_invitations(self) -> Any:
```
Возвращает list[AdministratorInvitationResponse].

ivr42 replied:
Да, но есть нюанс. Если поставить `async def get_all_invitations(self) -> list[AdministratorInvitationResponse]:`, то мой PyCharm и mypy ругается: "Expected type 'list[AdministratorInvitationResponse]', got 'list[AdministratorInvitation]' instead ".
(очевидно, что `list_all_invitations()` возвращает список из `AdministratorInvitation`)

В документации по fastapi это описано явно, и они рекомендуют использовать `Any`, если у нас жёсткая проверка типов. 

> If you have strict type checks in your editor, mypy, etc, you can declare the function return type as Any.
> 
> That way you tell the editor that you are intentionally returning anything. But FastAPI will still do the data documentation, validation, filtering, etc. with the response_model.

ADesBiysk replied:
Увидел. Можно оставить Any. 
Нужно будет использовать мапперы для преобразования всех моделей. Тогда от этой проблемы и некоторых можно будет избавиться. Но это позже.

ADesBiysk commented on src/core/db/repository/administrator_invitation.py:
```diff
@@ -17,9 +17,24 @@ def __init__(self, session: AsyncSession = Depends(get_session)) -> None:
+    async def get_all_invitations(self) -> list[AdministratorInvitation]:
+
+        result_statement = select(
+            AdministratorInvitation.name.label('name'),
```
Здесь можно забирать сразу всю модель. Лишние поля pydantic сам откинет, когда будет преобразовывать в AdministratorInvitationResponse. 
Соответственно, данный метод будет повторять task_repository.get_tasks (возможно, и другие где не запрашиваются связанные модели) и его можно будет перенести в AbstractRepository с именем list.
Поле expired_date некорректно названо в БД, поэтому правильнее его переименовать, чем постоянно переименовывать в уже загруженных данных.

ivr42 committed:
Code review fixes:

 - `responses` have been added to AdministratorInvitationCBV post and get methods
 - typing fix in `AdministratorInvitationCBV.get_all_invitations`
 - `expired_date` renamed to `expired_datetime` in `AdministratorInvitation` model
 - `async def get_all` added to `AbstractRepository`
 - `get_all_invitations` removed from `AdministratorInvitationRepository`, utilized `get_all` instead
 - `get_tasks` removed from `TaskRepository`, utilized `get_all` instead

ivr42 committed:
Merge branch 'develop' into feature/api_add_invitation_get_method

ADesBiysk commented on src/core/db/migrations/versions/2023-02-23_15.01.06_rename_administratorinvitation_expired_.py:
```diff
@@ -0,0 +1,30 @@
+
+# revision identifiers, used by Alembic.
+revision = 'ef4034f91b1c'
+down_revision = 'f4645f80c3ac'
```
Нужно загрузить последние изменения и сгенерировать миграцию заново. ID последней миграции сейчас fb0616f8ae8e. Из-за этого не проходит проверка в github workflow.

ADesBiysk commented on src/api/routers/administrator_invitation.py:
```diff
@@ -35,6 +42,17 @@ async def create_and_send_invitation(
+        response_description="Информация о приглашениях",
+    )
+    async def get_all_invitations(self) -> list[AdministratorInvitationResponse]:
+        return await self.administrator_invitation_service.list_all_invitations()
```
Этого метода не существует.

ivr42 replied:
да, к сожалению, при merge вылезли косяки, которы я не заметил глазами — поправил

ivr42 committed:
Fix migrations order

ivr42 committed:
Code-review fixes:

 - `expired-date` renamed to `expired-datetime`
 - method `list_all_invitations` in `AdministratorInvitationService`
   created
